File: skills/architecture/scripts/check-locality.mjs
import fs from 'node:fs';
import path from 'node:path';
import { extractImports, detectLang } from './utils/extractor.mjs';
import { findCycles } from './utils/graph.mjs';
import { walkDir } from './utils/walk.mjs';

/**
 * Resolve a relative import to candidate on-disk paths, per language.
 * The previous implementation only tried `.ts`/index.ts, so fan-out and cycle
 * detection silently returned empty for JS, Python, and Go projects.
 * @param {string} fromFile  File containing the import
 * @param {string} imp       Relative import specifier (starts with '.')
 * @param {'js'|'py'|'go'} lang
 * @returns {string[]}
 */
function importCandidates(fromFile, imp, lang) {
  const fromDir = path.dirname(fromFile);

  if (lang === 'py') {
    // Leading dots are package levels: one dot = current package (same dir),
    // each extra dot = one parent up.
    let dots = 0;
    while (imp[dots] === '.') dots++;
    const rest = imp.slice(dots).split('.').filter(Boolean);
    let base = fromDir;
    for (let d = 1; d < dots; d++) base = path.dirname(base);
    const resolved = path.join(base, ...rest);
    return [`${resolved}.py`, path.join(resolved, '__init__.py')];
  }

  // js / go (Go rarely uses relative file imports, but handle them gracefully)
  const resolved = path.resolve(fromDir, imp);
  const candidates = [];
  if (/\.[^./\\]+$/.test(path.basename(imp))) {
    // Import already carries an extension, e.g. './b.js' (ESM/Node style).
    candidates.push(resolved);
    // TS allows a '.js' specifier to resolve to the '.ts' source.
    const noExt = resolved.replace(/\.[^.]+$/, '');
    candidates.push(`${noExt}.ts`, `${noExt}.tsx`);
  }
  candidates.push(
    `${resolved}.ts`,
    `${resolved}.tsx`,
    `${resolved}.js`,
    `${resolved}.jsx`,
    `${resolved}.mjs`,
    `${resolved}.cjs`,
    `${resolved}.go`,
    path.join(resolved, 'index.ts'),
    path.join(resolved, 'index.tsx'),
    path.join(resolved, 'index.js'),
    path.join(resolved, 'index.jsx'),
    path.join(resolved, 'index.mjs'),
  );
  return candidates;
}

export function runLocalityCheck(
  targetDir,
  exclude = [
    'node_modules',
    '.test.',
    '.spec.',
    '.git',
    '.svn',
    '.hg',
    '.pytest_cache',
    '.tox',
    '__pycache__',
    '.venv',
    'venv',
    '.env',
    'dist',
    'build',
    'coverage',
    '.coverage',
    '.next',
    '.nuxt',
    '.cache',
    '.parcel',
    '.npm',
    '.yarn',
    'target',
    '.gradle',
    '.m2',
    '.pytest',
    '.mypy_cache',
    '.ruff_cache',
    '.vscode',
    '.idea',
    '.DS_Store',
  ],
) {
  const files = walkDir(targetDir, exclude);
  const graph = {};

  for (const file of files) {
    const content = fs.readFileSync(file, 'utf8');
    const lang = detectLang(file);
    const imports = extractImports(content, lang);
    graph[file] = [];

    for (const imp of imports) {
      // Only care about relative imports for locality
      if (!imp.startsWith('.')) continue;
      for (const candidate of importCandidates(file, imp, lang)) {
        try {
          fs.statSync(candidate);
          graph[file].push(candidate);
          break;
        } catch {
          // not found, try next candidate
        }
      }
    }
  }

  const cycles = findCycles(graph);

  // Calculate Fan-out
  const fanOut = Object.entries(graph)
    .map(([file, deps]) => ({ file, count: deps.length }))
    .sort((a, b) => b.count - a.count);

  return { cycles, fanOut };
}

// CLI entry point
if (
  process.argv[1] &&
  (process.argv[1] === new URL(import.meta.url).pathname ||
    process.argv[1].endsWith('check-locality.mjs'))
) {
  const dir = process.argv[2] || 'src';
  console.log(`Checking locality in ${dir}...`);
  try {
    const { cycles, fanOut } = runLocalityCheck(path.resolve(process.cwd(), dir));
    console.log('\n--- Circular Dependencies ---');
    cycles.forEach((cycle, i) => {
      console.log(`\nCycle ${i + 1}:`);
      cycle.forEach((c) => console.log(`  - ${path.relative(process.cwd(), c)}`));
    });
    if (cycles.length === 0) console.log('None found.');

    console.log('\n--- Top 5 Fan-out (Highest Imports) ---');
    fanOut.slice(0, 5).forEach((f) => {
      console.log(`  - ${path.relative(process.cwd(), f.file)} (${f.count} imports)`);
    });
  } catch (e) {
    if (e.code === 'ENOENT') {
      console.error(`Directory not found: ${dir}`);
    } else {
      console.error(e);
    }
  }
}

File: skills/architecting/scripts/check_locality.py
"""Module for checking code locality and detecting circular dependencies.

This script analyzes Python, JavaScript/TypeScript, and Go files to construct
an import graph, find circular dependencies, and compute file import fan-out.
"""

import argparse
import os
import sys
from pathlib import Path
from typing import Any

from utils.extractor import extract_imports, detect_lang
from utils.graph import find_cycles
from utils.walk import walk_dir, DEFAULT_EXCLUDE


def import_candidates(from_file: str, imp: str, lang: str) -> list[str]:
    """Generate possible file path candidates for a given relative import.

    Args:
        from_file: The path to the file containing the import.
        imp: The import specifier string (e.g., './module' or '..module').
        lang: The file language/extension ('py', 'js', 'go', etc.).

    Returns:
        A list of resolved candidate file paths to check for existence.
    """
    from_path = Path(from_file)
    from_dir = from_path.parent

    if lang == "py":
        # Leading dots are package levels: one dot = current package (same dir),
        # each extra dot = one parent up.
        dots = 0
        while dots < len(imp) and imp[dots] == ".":
            dots += 1

        if dots == 0:
            # Absolute import or not relative in the expected way
            # But locality check only cares about relative imports (starting with '.')
            return []

        rest = imp[dots:].split(".")
        rest = [r for r in rest if r]

        base = from_dir
        for _ in range(1, dots):
            base = base.parent

        resolved = base.joinpath(*rest)
        # ponytail: bare "from . import x" (rest=[]) only resolves via __init__.py;
        # PEP 420 namespace packages without __init__.py won't match. Fix if that
        # pattern shows up: capture the imported names and try them as submodules.
        return [f"{resolved}.py", str(resolved / "__init__.py")]

    # js / go
    resolved = (from_dir / imp).resolve()
    candidates = []

    basename = Path(imp).name
    if "." in basename:
        # Import already carries an extension
        candidates.append(str(resolved))
        # TS allows a '.js' specifier to resolve to the '.ts' source.
        no_ext = resolved.with_suffix("")
        candidates.extend([f"{no_ext}.ts", f"{no_ext}.tsx"])

    candidates.extend(
        [
            f"{resolved}.ts",
            f"{resolved}.tsx",
            f"{resolved}.js",
            f"{resolved}.jsx",
            f"{resolved}.mjs",
            f"{resolved}.cjs",
            f"{resolved}.go",
            str(resolved / "index.ts"),
            str(resolved / "index.tsx"),
            str(resolved / "index.js"),
            str(resolved / "index.jsx"),
            str(resolved / "index.mjs"),
        ]
    )
    return candidates


def run_locality_check(
    target_dir: str | Path, exclude: list[str] | None = None
) -> tuple[list[list[str]], list[dict[str, Any]]]:
    """Analyze a directory to find circular dependencies and top fan-out files.

    Args:
        target_dir: The directory to analyze.
        exclude: List of directory names or patterns to exclude.

    Returns:
        A tuple containing:
            - A list of cycles (each cycle is a list of file paths).
            - A list of dictionaries representing fan-out (highest imports).
    """
    if exclude is None:
        exclude = DEFAULT_EXCLUDE

    files = walk_dir(str(target_dir), exclude)
    graph = {}

    for file_path in files:
        try:
            content = Path(file_path).read_text(encoding="utf-8")
        except (UnicodeDecodeError, PermissionError):
            continue

        lang = detect_lang(file_path)
        imports = extract_imports(content, lang)
        graph[file_path] = []

        for imp in imports:
            # Only care about relative imports for locality
            if not imp.startswith("."):
                continue

            for candidate in import_candidates(file_path, imp, lang):
                if Path(candidate).exists():
                    graph[file_path].append(candidate)
                    break

    cycles = find_cycles(graph)

    # Calculate Fan-out
    fan_out = []
    for file, deps in graph.items():
        fan_out.append({"file": file, "count": len(deps)})

    fan_out.sort(key=lambda x: x["count"], reverse=True)

    return cycles, fan_out


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description="Check locality and circular dependencies."
    )
    parser.add_argument("dir", nargs="?", default="src", help="Directory to analyze")

    args = parser.parse_args()

    print(f"Checking locality in {args.dir}...")
    abs_dir = Path(args.dir).resolve()

    try:
        cycles, fan_out = run_locality_check(abs_dir)

        print("\n--- Circular Dependencies ---")
        if not cycles:
            print("None found.")
        else:
            for i, cycle in enumerate(cycles):
                print(f"\nCycle {i + 1}:")
                for node in cycle:
                    print(f"  - {os.path.relpath(node, Path.cwd())}")

        print("\n--- Top 5 Fan-out (Highest Imports) ---")
        for item in fan_out[:5]:
            print(
                f"  - {os.path.relpath(item['file'], Path.cwd())} ({item['count']} imports)"
            )

    except FileNotFoundError:
        print(f"Directory not found: {args.dir}", file=sys.stderr)
        sys.exit(1)
    except Exception as e:
        print(f"An error occurred: {e}", file=sys.stderr)
        sys.exit(1)
